### PR TITLE
Access Control

### PR DESCRIPTION
**Pull request by ba1ciu**

### CODE REVIEW

File: chain/app/app.go [hunk 1]
```diff
@@ -390,6 +396,7 @@ func New(
 		keys[plasticcreditmoduletypes.StoreKey],
 		keys[plasticcreditmoduletypes.MemStoreKey],
 		app.GetSubspace(plasticcreditmoduletypes.ModuleName),
+		app.GetPermStore(accesscontrolmoduletypes.ModuleName),
```
Comment (bjaanes) [inline chain/app/app.go:399]: This should send in the keeper, right?

File: chain/x/accesscontrol/types/perm_store.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
+package types
```
Comment (bjaanes) [inline chain/x/accesscontrol/types/perm_store.go:1]: This file makes me a little uncomfortable, because it deals with storage and stores and stuff outside of the keeper (go)module. I think it would be more correct to get this into the keeper in some way and send the keeper to the modules who are going to use this.

File: chain/x/accesscontrol/keeper/keeper.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+func (k Keeper) Logger(ctx sdk.Context) log.Logger {
+	return ctx.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName))
+}
+
+func (k Keeper) PermStore(name string) types.PermStore {
+	_, ok := k.permStores[name]
+	if ok {
+		panic("perm store already occupied")
```
Comment (bjaanes) [inline chain/x/accesscontrol/keeper/keeper.go:42]: More returning of errors and less panics, wtyd?

**Review by bjaanes — CHANGES_REQUESTED** [review]
Some thoughts on the design. I like the general concept, but I would prefer to keep it more in line with typical SDK design of passing keepers and using interfaces (expected keepers type of thing)

**Comment by bjaanes** [comment]
Oh, and spec documentation would be good :)

File: chain/x/accesscontrol/types/perm_store.go (hunk 2 above)
Comment (ba1ciu) [reply chain/x/accesscontrol/types/perm_store.go]: All storage op moved to keeper.

File: chain/x/accesscontrol/keeper/keeper.go (hunk 3 above)
Comment (ba1ciu) [reply chain/x/accesscontrol/keeper/keeper.go]: Moved to subkeeper's initSubStore(). I'd personally leave it like that as SubKeeper is ought to be initialized only during app initialization in app.go

File: chain/app/app.go (hunk 1 above)
Comment (ba1ciu) [reply chain/app/app.go]: Now SubKeeper is initialized and injected.

File: chain/app/test_helpers.go [hunk 4]
```diff
@@ -181,3 +182,10 @@ func CreateRandomAccounts(accNum int) []sdk.AccAddress {
 
 	return testAddrs
 }
+
+func InitAccessControlSubKeepers(keeper *accesscontrolmodulekeeper.Keeper) *[]accesscontrolmodulekeeper.IAccessControlSubKeeper {
+	var sk []accesscontrolmodulekeeper.IAccessControlSubKeeper
+	sk = append(sk, accesscontrolmodulekeeper.NewSubKeeper(keeper, "mockmodule1"))
```
Comment (gjermundgaraba) [inline chain/app/test_helpers.go:188]: Do they have to be initialized?
I'm just wondering if it is less error-prone if they are just lazily initialized?

File: chain/proto/empowerchain/accesscontrol/query.proto [hunk 5]
```diff
@@ -0,0 +1,11 @@
+syntax = "proto3";
+package empowerchain.accesscontrol;
+
+import "gogoproto/gogo.proto";
+import "google/api/annotations.proto";
+
+option go_package = "github.com/empowerchain/empowerchain/x/accesscontrol/types";
+
+service Query {
```
Comment (gjermundgaraba) [inline chain/proto/empowerchain/accesscontrol/query.proto:9]: Would it make more sense to just not have these empty proto files if they are not really needed?

File: chain/x/accesscontrol/keeper/keeper.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+func (k Keeper) revokeAccess(ctx sdk.Context, subKey string, account sdk.AccAddress, msgType string) {
+	store := k.kvStore(ctx, subKey)
+	key := types.PermissionStoreKey(account, msgType)
+
+	store.Delete(key)
+}
+
+// Should be used only for genesis export!
+func (k Keeper) IteratePermissions(ctx sdk.Context, subKey string,
```
Comment (gjermundgaraba) [inline chain/x/accesscontrol/keeper/keeper.go:67]: If this is only for genesis export, can it not be private?

File: chain/x/accesscontrol/keeper/sub_keeper.go [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+type IAccessControlSubKeeper interface {
+	HasAccess(ctx sdk.Context, account sdk.AccAddress, msgType string) bool
+	GrantAccess(ctx sdk.Context, account sdk.AccAddress, msgType string) error
+	RevokeAccess(ctx sdk.Context, account sdk.AccAddress, msgType string)
+}
+
+type SubKeeper struct {
+	*Keeper
+	subKey string
+	kaddr  string
```
Comment (gjermundgaraba) [inline chain/x/accesscontrol/keeper/sub_keeper.go:14]: These names are very hard to parse and understand. Perhaps it might be better to give them more descriptive names?

File: chain/x/accesscontrol/module.go [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+
+func (AppModuleBasic) RegisterLegacyAminoCodec(*codec.LegacyAmino) {}
+
+// RegisterInterfaces registers the module's interface types
+func (AppModuleBasic) RegisterInterfaces(registry codectypes.InterfaceRegistry) {
+	types.RegisterInterfaces(registry)
+}
+
+func (AppModuleBasic) DefaultGenesis(cdc codec.JSONCodec) json.RawMessage {
+	return nil
```
Comment (gjermundgaraba) [inline chain/x/accesscontrol/module.go:58]: Is nil valid here?

File: chain/x/accesscontrol/types/keys.go [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/address"
+	"github.com/cosmos/cosmos-sdk/types/kv"
+)
+
+const (
+	// ModuleName defines the module name
+	ModuleName = "accesscontrol"
+
+	// Workaround for key collision
```
Comment (gjermundgaraba) [inline chain/x/accesscontrol/types/keys.go:15]: Can you explain this a bit more?

**Review by gjermundgaraba — CHANGES_REQUESTED** [review]
Looks good, mostly just some questions we can go through on Monday

File: chain/proto/empowerchain/accesscontrol/query.proto (hunk 5 above)
Comment (ba1ciu) [reply chain/proto/empowerchain/accesscontrol/query.proto]: Yes. I deleted protos and querier + msg server.

File: chain/x/accesscontrol/keeper/keeper.go (hunk 6 above)
Comment (ba1ciu) [reply chain/x/accesscontrol/keeper/keeper.go]: Fixed

File: chain/x/accesscontrol/keeper/sub_keeper.go (hunk 7 above)
Comment (ba1ciu) [reply chain/x/accesscontrol/keeper/sub_keeper.go]: Leftover from debugging. Deleted.

File: chain/x/accesscontrol/module.go (hunk 8 above)
Comment (ba1ciu) [reply chain/x/accesscontrol/module.go]: Fixed.

File: chain/x/accesscontrol/types/keys.go (hunk 9 above)
Comment (ba1ciu) [reply chain/x/accesscontrol/types/keys.go:15]: Added a comment.

**Review by gjermundgaraba — CHANGES_REQUESTED** [review]
Conflicts that needs to be solved, but otherwise good
